journalkeeper-core/src/main/java/io/journalkeeper/core/server/VoterConfigManager.java

```diff
@@ -83,11 +77,21 @@ boolean maybeUpdateLeaderConfig(UpdateRequest request,
                 waitingForAllEntriesCommitted(journal);
                 votersConfigStateMachine.toNewConfig(appendEntryCallable);
                 Collection<URI> followerUris = leader.getFollowerUris();
+                // 如果有follower被删除，需要等待新配置都被提交
```

Comment:
appendEntryCallable 里将newConfig 写入了日志， 如果此时不等待这条日志被commit, 直接移除follower 会导致被移除的节点不知道自己被移除，一直向旧voters 发送选举请求

Reply:
这个逻辑是没有必要的，Raft通过 “心跳时间间隔 << 心跳超时” 来解决这个问题。参见Raft论文：

The third issue is that removed servers (those not in
Cnew) can disrupt the cluster. These servers will not receive heartbeats, so they will time out and start new elections. They will then send RequestVote RPCs with new
term numbers, and this will cause the current leader to
revert to follower state. A new leader will eventually be
elected, but the removed servers will time out again and
the process will repeat, resulting in poor availability.

To prevent this problem, servers disregard RequestVote
RPCs when they believe a current leader exists. Specifically, if a server receives a RequestVote RPC within
the minimum election timeout of hearing from a current leader, it does not update its term or grant its vote.
This does not affect normal elections, where each server
waits at least a minimum election timeout before starting
an election. However, it helps avoid disruptions from removed servers: if a leader is able to get heartbeats to its
cluster, then it will not be deposed by larger term numbers.

Reply:
新的集群不会因为被移除的节点的new election扰乱。但是被移除的节点一直没有停下来(没有收到配置变更第二阶段的日志)，不停地在发送new election请求. 如果有需要，我可以再重现一下这个场景， 补充上异常日志

Reply:
是的，目前不会自动停止被移出节点。

Reply:
# 二阶段配置变更

Raft论文中，给出了一种安全的二阶段配置变更方法，JournalKeeper在实现过程中，对这个方法做了一些改进。

## 为什么不采用Raft论文中的实现方法

Raft协议中，2阶段配置变更的过程，需要一个临时的、中心化的协调者，一般来说是发起变更的客户端来担任。

为了保证各种故障的场景下，尽可能的完成二阶段配置变更，我们需要一个持久化分布式存储来保存中间状态信息。考虑到JournalKeeper本身就是一个分布式协调服务，我们不希望增加一个外部依赖。但如果自行实现难度较大，变更失败的中间状态回滚也很难处理。

## JournalKeeper做了哪些改变？带来了什么问题？

JournalKeeper解决方案是，使用当前的Leader作为协调者，使用保留分区的日志InternalEntry来作为持久化存储，持久化变更日志，实现自我协调的配置变更。

这种实现的一个副作用是，集群多次配置变更的日志，有可能会被新加入的节点在状态机中回放。虽然节点在回放旧的配置变更日志时，会临时改变自己的配置，但只要节点把所有日志都回放完成，最终会变为最新配置。所以某个节点在处理配置变更时，即使新配置中没有自己，也不能把自己停掉。

比如，新加入集群的节点D，它需要从某一个Snapshot开始，在状态机中回放之后的所有日志，以便让节点的状态数据和集群保持一致。

旧的配置变更日志也会被回放，例如：

第一次配置变更：A -> A, B, C
第二次配置变更:  A, B, C -> A, B, D

当启动节点D并加入集群后， 当节点D状态机执行第一次配置变更的日志A -> A, B, C的时候，不能因为新配置中没有自己就自动把自己停掉，否则它就无法继续回放其它日志，最终变成最新配置了。

## 如何解决被移出集群的节点无法自动停止的问题？

考虑到被移出集群的节点不能自动停止，可以由发起配置变更的客户端订阅ON_VOTERS_CHANGE事件，待集群完成配置变更之后，调用RaftServer##stop方法停止节点。